### PR TITLE
[Lib][ML] Linear Regression, Logistic Regression & SVM

### LINKED ISSUE #89: Add Linear Regression, Logistic Regression and SVM Implementation
To implement Linear Regression, Logistic Regression and SVM using new husky API.

### PR DESCRIPTION
**Pull request by TatianaJin**
Resolve #89.
1. Add FeatureLabelBase to abstract data objects
2. Add DataLoader to read LIBSVM/TSV format from HDFS as data objects
3. Add Regression to abstract linear and logistic regression models
4. Add GradientDescentBase to abstract SGD and FGD
5. Add ParameterBase to abstract ML model parameters

### CODE REVIEW

File: lib/sgd.hpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    void update_vec(ObjL& data, ParamT param_list, int num_global_samples) override {
+        ASSERT_MSG(this->learning_rate_ != 0, "Learning rate is set to 0.");
+        ASSERT_MSG(this->gradient_func_ != nullptr, "Gradient function is not specified.");
+
+        int num_local_samples = data.get_size();              // number of local samples
+        vec_double current_vec = param_list.get_all_param();  // local copy of parameter
+        auto& ac = AggregatorFactory::get_channel();
+
+        list_execute(data, {}, {&ac}, [&grad_func = this->gradient_func_, &rate = this->learning_rate_,
```
Comment (ddmbr) [inline lib/sgd.hpp]: This looks like FGD?

Comment (TatianaJin) [inline lib/sgd.hpp]: The only difference is that for FGD current_vec is not changed in list_execute but it is changed for SGD... What is your suggestion?

**Comment by kygx-legend** [comment]
Please create a subdir `ml` in `lib` and use `husky::lib::ml` namespace instead.

File: lib/sgd.hpp (hunk 1 above)
Comment (ddmbr) [inline lib/sgd.hpp]: My suggestion is that you may just leave it for now.

Previously we had a list_execute that accept a Sequencer as one of the arguments. A Sequencer is to generate a series of indices, in order to instruct the execution of objects at different indices.

With the use of Sequencer, we can easily create a list_execute that can sample some objects to execute. Maybe you can continue to work on this and eventually come up with a better SGD, after this PR?

**Comment by ddmbr** [comment]
I agree with @kygx-legend . This makes the folder more organized.

**Comment by ddmbr** [comment]
Ping @TatianaJin . Please organize the files as suggested by @kygx-legend .

**Comment by TatianaJin** [comment]
@ddmbr @kygx-legend I see. Thank you for your advice.

**Comment by ddmbr** [comment]
As is required by @kygx-legend , please move these files into a new folder called `ml/`, i.e., `lib/ml/`.

**Comment by Kelvin-Ng** [comment]
Now my Vector and DataPoint classes are merged. It would be better to change the patch to use my Vector and DataPoint. Also note that your algorithms should support both dense and sparse input by making use of the high level APIs of my Vector class (in particular the iterators). It should be good enough so that you can achieve 'write once, done for both dense and sparse'. If you need any function that is not support by my Vector class, feel free to tell me and I will add it if possible.

Examples of using it can be found at https://github.com/Kelvin-Ng/VR-SGD-Implementation. It aims at providing a collection of variants of SGD implemented for *single-machine single-thread* situation. In particular, note that in algo/saga.hpp, template parameter `is_sparse` is used to allow users to switch between dense and sparse input. In lib/prox.hpp, iterator is used so that `prox_l1()` can be used for both dense and sparse vector. [The old version of `prox_l1()` without using iterator](https://github.com/Kelvin-Ng/VR-SGD-Implementation/blob/11e2698cc86cb8f996fa1f55ec27caffe9da206f/lib/prox.hpp) is very clumsy.

**Comment by TatianaJin** [comment]
@Kelvin-Ng Thanks for your comment.
Is it we suppose users always use Vector/DataPoint on husky to store parameters and data points? I mean, if there is a need for some extension on data points, say data points for random forest that maintain a state for occurrences in tree nodes, does a new class have to be implemented inheriting from DataPoint? 

I really appreciate your 'write once, done for both dense and sparse' philosophy. I have previously thought about it when I worked on updates to parameters, but got stuck in how to implement appropriate iterators, I would like to discuss with you later about uniform operations after studying your codes. : )

**Comment by Kelvin-Ng** [comment]
@TatianaJin Yes, users should always store their data inside Vector and DataPoint **or their subclasses**. Users are free to extend them by inheriting them. Some algorithms like random forest may need an extension of DataPoint, and it is fine to ask the users to store their data in that extension because that subclass of DataPoint can still be fed into other algorithms requiring standard DataPoint.

**Comment by kygx-legend** [comment]
This PR holds for a long time, which is not efficient. Could you separate this into 3 smaller tasks such as `Support Linear Regression`, `Support Logistic Regression` and `Support SVM`? It would reduce the cost of review and modification.

**Comment by ddmbr** [comment]
@kygx-legend How about let @TatianaJin  rebase and then we merge this one first, and then do more fixes later? I think some working examples are the highest priority. We can deal other things with more follow-up PRs.

**Comment by kygx-legend** [comment]
@ddmbr , if this PR does not block other parts, I think it's good. Please speed up.

**Comment by ddmbr** [comment]
Ping @TatianaJin . Please rebase and resolve conflicts.

**Comment by kygx-legend** [comment]
I'll merge this and please follow up those parts which are not ready.

**Comment by ddmbr** [comment]
@rylanchiu Notice that this PR is now in master and you may want to use some of the new components.

**Pull request merged**